### PR TITLE
feat: user v3 contact email and phone

### PR DESCRIPTION
**Pull request by stebenz**
# Which Problems Are Solved

Endpoints to maintain email and phone contact on user v3 are not implemented.

# How the Problems Are Solved

Add 3 endpoints with SetContactEmail, VerifyContactEmail and ResendContactEmailCode.
Add 3 endpoints with SetContactPhone, VerifyContactPhone and ResendContactPhoneCode.
Refactor the logic how contact is managed in the user creation and update.

# Additional Changes

None

# Additional Context

- part of https://github.com/zitadel/zitadel/issues/6433

### CODE REVIEW

**Review by livio-a — CHANGES_REQUESTED** [review]
the PR title and description need to update to reflect the phone part as well

File: internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+			res: res{
+				want: &resource_object.Details{
+					Changed: timestamppb.Now(),
+					Owner: &object.Owner{
+						Type: object.OwnerType_OWNER_TYPE_ORG,
+						Id:   orgResp.GetOrganizationId(),
+					},
+				},
+			},
+		},
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go:158]: True, missed that, as I was replacing the emails with gofakeit.Email(), corrected

File: internal/api/grpc/resources/user/v3alpha/user.go [hunk 2]
```diff
@@ -22,12 +21,12 @@ func (s *Server) CreateUser(ctx context.Context, req *user.CreateUserRequest) (_
 	if err != nil {
 		return nil, err
 	}
-
-	if err := s.command.CreateSchemaUser(ctx, schemauser, s.userCodeAlg); err != nil {
+	details, err := s.command.CreateSchemaUser(ctx, schemauser, s.userCodeAlg)
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/user.go:24]: Done

File: internal/api/grpc/resources/user/v3alpha/email.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	object "github.com/zitadel/zitadel/pkg/grpc/object/v3alpha"
+	user "github.com/zitadel/zitadel/pkg/grpc/resources/user/v3alpha"
+)
+
+func (s *Server) SetContactEmail(ctx context.Context, req *user.SetContactEmailRequest) (_ *user.SetContactEmailResponse, err error) {
+	if err := checkUserSchemaEnabled(ctx); err != nil {
+		return nil, err
+	}
+	schemauser := setContactEmailRequestToChangeSchemaUserEmail(req)
+	details, err := s.command.ChangeSchemaUserEmail(ctx, schemauser, s.userCodeAlg)
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/email.go:20]: Done

File: internal/api/grpc/resources/user/v3alpha/email.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		Verified:    setEmail.GetIsVerified(),
+		URLTemplate: setEmail.GetSendCode().GetUrlTemplate(),
+	}
+}
+
+func (s *Server) VerifyContactEmail(ctx context.Context, req *user.VerifyContactEmailRequest) (_ *user.VerifyContactEmailResponse, err error) {
+	if err := checkUserSchemaEnabled(ctx); err != nil {
+		return nil, err
+	}
+	details, err := s.command.VerifySchemaUserEmail(ctx, organizationToUpdateResourceOwner(req.Organization), req.GetId(), req.GetVerificationCode(), s.userCodeAlg)
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/email.go:54]: Done

File: internal/api/grpc/resources/user/v3alpha/email.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	return &user.SetContactEmailResponse{
+		Details:          resource_object.DomainToDetailsPb(details, object.OwnerType_OWNER_TYPE_ORG, details.ResourceOwner),
+		VerificationCode: gu.Ptr(schemauser.ReturnCode),
+	}, nil
+}
+
+func setContactEmailRequestToChangeSchemaUserEmail(req *user.SetContactEmailRequest) *command.ChangeSchemaUserEmail {
+	return &command.ChangeSchemaUserEmail{
+		ResourceOwner: organizationToUpdateResourceOwner(req.Organization),
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/email.go:32]: If you want to change a user in another context than your own, for example as a System user or in a specific org

File: internal/api/grpc/resources/user/v3alpha/email.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		return nil, err
+	}
+	schemauser := setContactEmailRequestToChangeSchemaUserEmail(req)
+	details, err := s.command.ChangeSchemaUserEmail(ctx, schemauser, s.userCodeAlg)
+	if err != nil {
+		return nil, err
+	}
+	return &user.SetContactEmailResponse{
+		Details:          resource_object.DomainToDetailsPb(details, object.OwnerType_OWNER_TYPE_ORG, details.ResourceOwner),
+		VerificationCode: gu.Ptr(schemauser.ReturnCode),
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/email.go:26]: Done

File: internal/command/email.go [hunk 7]
```diff
@@ -29,3 +29,16 @@ func (e *Email) Validate() error {
...
+}
+func (r *Email) IsVerified() bool {
+	return r.Verified
+}
+func (r *Email) IsReturnCode() bool {
+	return r.ReturnCode
+}
+func (r *Email) GetURLTemplate() string {
+	return r.URLTemplate
+}
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/email.go:44]: Done

File: internal/command/command.go [hunk 8]
```diff
@@ -189,13 +189,37 @@ type AppendReducer interface {
...
+	if len(cmds) == 0 {
+		return writeModelToObjectDetails(object.GetWriteModel()), nil
+	}
+	events, err := c.eventstore.Push(ctx, cmds...)
+	if err != nil {
+		return nil, err
+	}
+	if err := AppendAndReduce(object, events...); err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/command.go:219]: Done

File: internal/command/command.go [hunk 9]
```diff
@@ -189,13 +189,37 @@ type AppendReducer interface {
...
 	}
 	return AppendAndReduce(object, events...)
 }
 
+type WriteModel interface {
+	AppendEvents(...eventstore.Event)
+	// TODO: Why is it allowed to return an error here?
+	Reduce() error
+	GetWriteModel() *eventstore.WriteModel
+}
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/command.go:207]: Done

File: internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+			got, err := instance.Client.UserV3Alpha.ResendContactEmailCode(tt.ctx, tt.req)
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+			integration.AssertResourceDetails(t, tt.res.want, got.Details)
+			if tt.res.returnCode {
+				require.NotNil(t, got.VerificationCode)
+			}
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go:693]: Done

File: internal/api/grpc/resources/user/v3alpha/phone.go
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/phone.go:1]: Done

File: internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go [hunk 11]
```diff
@@ -0,0 +1,696 @@
...
+				"type": "string"
+			}
+		}
+	}`)
+	schemaResp := instance.CreateUserSchema(isolatedIAMOwnerCTX, schema)
+	orgResp := instance.CreateOrganization(isolatedIAMOwnerCTX, gofakeit.Name(), gofakeit.Email())
+
+	type res struct {
+		want            *resource_object.Details
+		returnCodeEmail bool
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go:311]: Done

File: internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go [hunk 12]
```diff
@@ -0,0 +1,696 @@
...
+			got, err := instance.Client.UserV3Alpha.SetContactEmail(tt.ctx, tt.req)
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+			integration.AssertResourceDetails(t, tt.res.want, got.Details)
+			if tt.res.returnCode {
+				require.NotNil(t, got.VerificationCode)
+			}
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go:286]: Done

File: internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go [hunk 13]
```diff
@@ -0,0 +1,696 @@
...
+				want: &resource_object.Details{
+					Changed: timestamppb.Now(),
+					Owner: &object.Owner{
+						Type: object.OwnerType_OWNER_TYPE_ORG,
+						Id:   orgResp.GetOrganizationId(),
+					},
+				},
+			},
+		},
+	}
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/integration_test/email_test.go:270]: Done

File: internal/api/grpc/resources/user/v3alpha/integration_test/phone_test.go
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/grpc/resources/user/v3alpha/integration_test/phone_test.go:1]: No, changed

**Review by livio-a — APPROVED** [review]

**Pull request merged**